### ISSUE TITLE
Perf issue: Torchsharp is slower than pytorch on cuda on some operators

### ISSUE THREAD

**Issue by LittleLittleCloud** [dotnet/TorchSharp#1442]
I run some benchmark tests to compare the performance difference between torchsharp and pytorch, both uses `libtorch 2.2.1 + cuda 12.1`. And I notice that torchsharp is slower than pytorch in most of operators. Below are benchmark result

## Torchsharp

![Image](https://github.com/user-attachments/assets/75927d6d-8e35-4ea8-bef8-c808a02957e8)

## Pytorch

![Image](https://github.com/user-attachments/assets/14118081-9215-4b26-ae21-50449ee8d19c)

## Source code
```csharp
using TorchSharp;

// Initialize CUDA device
var device = torch.CUDA;

var repeatTime = 10000;
// Test randn
var startTime = DateTime.Now;
for (int i = 0; i < repeatTime; i++)
{
    var _ = torch.randn(new long[] { 1000, 1000 }, device: device);
}

Console.WriteLine("Time taken for randn: " + (DateTime.Now - startTime).TotalSeconds);

// Test matmul
var a = torch.randn(new long[] { 1000, 1000 }, device: device);
var b = torch.randn(new long[] { 1000, 1000 }, device: device);

for (int i = 0; i < repeatTime; i++)
{
    var c = torch.matmul(a, b);
}

Console.WriteLine("Time taken for matmul: " + (DateTime.Now - startTime).TotalSeconds);

// Test concat
startTime = DateTime.Now;
a = torch.randn(new long[] { 1000, 1000 }, device: device);
b = torch.randn(new long[] { 1000, 1000 }, device: device);

for (int i = 0; i < repeatTime; i++)
{
    var c = torch.cat(new[] { a, b }, 0);
}

Console.WriteLine("Time taken for concat: " + (DateTime.Now - startTime).TotalSeconds);

// Test slice
startTime = DateTime.Now;
a = torch.randn(new long[] { 1000, 1000 }, device: device);

for (int i = 0; i < repeatTime; i++)
{
    var c = a[.., 0..500];
}

Console.WriteLine("Time taken for slice: " + (DateTime.Now - startTime).TotalSeconds);

// Test add
startTime = DateTime.Now;
a = torch.randn(new long[] { 1000, 1000 }, device: device);
b = torch.randn(new long[] { 1000, 1000 }, device: device);

for (int i = 0; i < repeatTime; i++)
{
    var c = a + b;
}

Console.WriteLine("Time taken for add: " + (DateTime.Now - startTime).TotalSeconds);
```

```python
# create a list of benchmark for pytorch on cuda

import torch
import time
repeat = 10000
total_time = 0
start_time = time.time()
for _ in range(repeat):
    a = torch.randn(1000, 1000).cuda()
print("Time taken for randn: " , time.time()-start_time)

start_time = time.time()
# test matmul
a = torch.randn(1000, 1000).cuda()
b = torch.randn(1000, 1000).cuda()
for _ in range(repeat):
    c = torch.matmul(a, b)
    

print("Time taken for matmul: ", time.time()-start_time)

start_time = time.time()

# test concat   
a = torch.randn(1000, 1000).cuda()
b = torch.randn(1000, 1000).cuda()

for _ in range(repeat):
    c = torch.cat((a, b), 0)

print("Time taken for concat: ", time.time()-start_time)

start_time = time.time()
# test slice
a = torch.randn(1000, 1000).cuda()

for _ in range(repeat):
    c = a[:, 0:500]

print("Time taken for slice: ", time.time()-start_time)

start_time = time.time()
# test add
a = torch.randn(1000, 1000).cuda()
b = torch.randn(1000, 1000).cuda()

for _ in range(repeat):
    c = a + b

print("Time taken for add: ", time.time()-start_time)
```

**Comment by ds5678** [comment]
For something like this, you should use [BenchmarkDotNet](https://github.com/dotnet/BenchmarkDotNet). It handles all the edge cases with .NET benchmarking (like JIT warmup).

**Comment by ds5678** [comment]
@ozanMSFT I think you meant to ping @LittleLittleCloud

**Comment by haytham2597** [comment]
Is interesting the performance that you have in matmul so i replicate them with Release mode in TorchSharp and LibTorch 2.8.0 cu128.

As always because of cache the first time may take some time, all times is represent in miliseconds.

<img width="766" height="648" alt="Image" src="https://github.com/user-attachments/assets/d4318408-13f2-4a02-8dc8-daf527f5eae1" />

The source code that i do is:

C#:
```cs
var device = torch.CUDA;

var repeatTime = 1000;
long[] dims = new long[] { 100, 100 };
for (int n = 0; n < 10; n++)
{
    Console.WriteLine($"{n+1}/10");
    torch.backends.cudnn.allow_tf32 = true;
    torch.backends.cuda.matmul.allow_tf32 = true;
    var startTime = Stopwatch.GetTimestamp();
    using (var a = torch.randn(dims, device: device, dtype: torch.ScalarType.Float32))
    using (var b = torch.randn(dims, device: device, dtype: torch.ScalarType.Float32))
    {
        for (int i = 0; i < repeatTime; i++)
        {
            var c = torch.matmul(a, b);
            c.Dispose();
        }
        Console.WriteLine($"Time taken for matmul: {new TimeSpan(Stopwatch.GetTimestamp() - startTime).TotalMilliseconds}ms");
        GC.Collect();
        Thread.Sleep(1000);
        Console.WriteLine("Now test with bfloat16");

        startTime = Stopwatch.GetTimestamp();
        a.to(torch.ScalarType.BFloat16);
        b.to(torch.ScalarType.BFloat16);

        for (int i = 0; i < repeatTime; i++)
        {
            var c = torch.matmul(a, b);
            c.Dispose();

        }

        Console.WriteLine($"Time taken for matmul bf16: {new TimeSpan(Stopwatch.GetTimestamp() - startTime).TotalMilliseconds}ms");
    }
}
```

Pytorch
```python

import torch
import time
from datetime import datetime
import gc

repeat = 1000
dims =100
total_time = 0

def to_ms(difftime):
    return difftime.total_seconds()*1000

for n in range(10):
    print(f"{n+1}/10")
    start_time = datetime.now()
    
    torch.backends.cuda.matmul.allow_tf32 = True
    torch.backends.cudnn.allow_tf32 = True
    
    a = torch.randn(dims,dims).cuda()
    b = torch.randn(dims,dims).cuda()
    for _ in range(repeat):
        c = torch.matmul(a, b)
    
    
    print("Time taken for matmul: ", to_ms(datetime.now()-start_time))
    gc.collect()
    print("Now test with bfloat16")
    time.sleep(1)
    start_time = datetime.now()
    
    
    a = a.to(dtype=torch.bfloat16)
    b = b.to(dtype=torch.bfloat16)
    
    for _ in range(repeat):
        c = torch.matmul(a, b)
    
    print("Time taken for matmul bf16: ", to_ms(datetime.now()-start_time))
```

FAQ:
Why in C# i use the Using statement on tensors?
because i have a theory that in pytorch automatically dispose the tensor outer-scope (warning i'm not sure. I didn't researched that)

Why sometime in c# torchsharp the BFloat16 is slow than F32?
I don't know i try figure out, i have some theory, but not pretty sure. Need more test.

Specification?
- i7 9700Kf
- RTX 3070 
- 32Gb Ram

What version torch between TorchSharp and Pytorch used?
| C#  | Python |
| ------------- | ------------- |
| LibTorch 2.8.0 Cu128| 2.7.1+cu126 |

**Comment by haytham2597** [comment]
I does more test about matmul and notice the difference between TorchSharp and Pytorch is around of ~22 microseconds (0.022 miliseconds) in favor of Pytorch and at least i know the reason is about of DllImport the system take a time to call a function of Dll. 

In LibTorch C++ the diffeence is around of ~5 microseconds (0.005 miliseconds) in favor of C++

In resume, is imperceptible the speed between TorchSharp and Pytorch.

Soon i will make a table for get better comparison between TorcSharp, Pytorch and LibTorch C++ and with different environment of Net

All that test is under BFloat16 and Net Framework 4.7.2
